### DISCUSSION TITLE
[feature] Retraining a fitted model

### DISCUSSION THREAD

**Discussion by Msaleh87** [ourownstory/neural_prophet#729] [Feature Requests & Feedback]
I know how to do this in fbProphet, you get the parameters from the old model and update them using the new data points, instead of retraining the whole model from scratch. How to do the same with Neural Prophet? I tried the same function but it didn't work out.

**Comment by ourownstory** [comment]
Hi @Msaleh87 
We currently have disabled retraining of an already fitted model.
Do you mind sharing how this would be beneficial to your application?

**Reply by Msaleh87** [reply]
Hi @ourownstory 
Thanks for getting back to me on this, appreciate it!
So currently I am working on a data-driven Microgrid controller. A microgrid is a miniaturized version of the electric grid, it has solar panels, batteries and is connected to the main electric grid. Now, the controller of the microgrid is trying to optimize its operation, meaning, relaying as less as possible on the main electric grid and using the solar panels and batteries (or whatever assets it has). Also, it tries to optimize the cost operation, during high charges from the grid, it would use its own resources. However, in order for the controller to do good optimization, it needs accurate forecast of what is going to happen, so the controller takes the forecast of the 48-hours ahead of the electric loads and solar panels and the rest of the resources. Now, for forecasting, I train my model and it's working fine, but the real-time measurements keep accumulating and I could use it to retrain the fitted model to get a more accurate forecast, especially if the historical data were less to begin with. Please let me know if any of this is not clear.

**Reply by Msaleh87** [reply]
I think if this feature is enabled, it would be a huge asset in the electric smart grid applications, which have real-time measurement happening constantly (like every 1-5 mins). I think any application that has real-time measurements could utilize this.

**Reply by ourownstory** [reply]
Hi @Msaleh87 
That makes sense. It seems like there are two things that you need to update to get accurate forecasts: 
a) the state of the system / input data
b) the model parameters
Unless you have very little data, I would assume that (a) changes significantly more than (b) within a short time span. 
Thus, you could update (a) your input data (your prediction DataFrame), and get new predictions reflecting the updated state every 1-5 minutes. Meanwhile, when you have collected enough new data, you can retrain (b) the whole model once every day or week. 

Online learning algorithms do exist, but they often have different trade-offs. 

hope this helps

**Reply by Msaleh87** [reply]
It was a great help just telling me that the retraining was disabled, it saved me a lot of searching time, thanks!

So I have an electric load data for about a year, I trained the model for a data set from Mar 2021 until Mar 2022. I enabled the holidays. Here is my code:

```
m = NeuralProphet(n_forecasts=(2*24*12),
    n_lags=(16*24*12),
    yearly_seasonality="auto",
    weekly_seasonality="auto",  
    daily_seasonality="auto", 
    growth="off", seasonality_reg= 0.5, optimizer = 'AdamW',
    learning_rate = 0.1, n_changepoints="auto", changepoints_range = "auto" ) # trend_reg=1
m.add_country_holidays("US", mode="additive") # lower_window=-1, upper_window=1
metrics = m.fit(L12M, freq="5min")

m.highlight_nth_step_ahead_of_each_forecast(step_number=m.n_forecasts)

P_horizon = 576 # 2 days ahead
future = m.make_future_dataframe(df = L12M, periods = P_horizon, n_historic_predictions=576)
forecast1 = m.predict(df=future)
```
That's the 48 hours ahead I got, I was wondering what hyperparamters I could add/tune to have better results. Or this is as good as it gets since I am using 1 feature (holidays)?

![Figure 2 (4)](https://user-images.githubusercontent.com/42920113/186675589-f5f8a7d0-3da7-4156-9182-6da9b478a299.png)

**Reply by ourownstory** [reply]
Hi @Msaleh87 Do you mind creating a new Question in the forecasting help section?
some pointers: 
* Try a lower amount of lags (e.g. just 3-7 days). OR downsample your data to 15min or hourly OR add a hidden layer of a small size (e.g. 8-32) (reason: your current AR model is massive, it has 2\*24\*12 \* 16\*24\*12 = 2,654,208 parameters, and is likely overfitting)
* instead of regularizing seasonality, try setting the number of fourier terms (instead of "auto")
* use the plot_components and plot_parameters function and see if the make sense
* find external regressors like temp
* at the end, tune learning rate and epochs

**Reply by Msaleh87** [reply]
I just did, here is the link: https://github.com/ourownstory/neural_prophet/discussions/741
So I need to predict 48 hours ahead, which is 576 data points (2 days* 48 hrs*12 datapoints per hr), so I used n_forecast = 576. as for n_lags I tries to look back 2-3 days, but the forecast used to come out bad. When I played with the hidden layers and neurons, it was also not that great, even though I think I used 3 hidden layers and 64 neurons I guess. I also tried to set seasonality like the documentation daily is set to a default of 6 and weekly to 4, in the documentation it mentioned that I should use multiple of those (like for daily, it should be m * 6, and for weekly: n * 4, where m and n are integers) to capture the complex seasonality. I guess I'll try to change seasonality again, but I remember that I did, but I'll give it another try. But I am sure that n_lags that big started to give me reasonable forecasts like the one in the picture. I know since the horizon is big I have to sacrifice some of the accuracy, but I was hoping for more accurate model, or better tune it.
Is there a way, besides grid search to tune a Neural prophet model?

**Reply by ourownstory** [reply]
replied: https://github.com/ourownstory/neural_prophet/discussions/741#discussioncomment-3487281

**Reply by tercjak** [reply]
I think the best is to enable updating data to trained model.
When we have this feature we could train 3 models (n_forecast1=3 A,n_forecast2,=4 B, n_forecast3=5 C) so we can get forecasts with lower amount of steps (eg 3-5 -1 step 6-10 2steps 11-15 3steps etc.  )   3,4,5  6,7,8 ,9..= A,B,C,AA,AB,AC,BB,BC,CC ... etc. For now the only option to do this is to add lagged Y data as regressor. There is no need to sacrifice accuracy if the horizon is larger; on the contrary, a large amount of data should give us a more accurate model. If the opposite is the case, our methodology is flawed.
